### ISSUE TITLE
me.Renderable.getBounds() returns a relative rect

### ISSUE THREAD

**Issue by parasyte** [melonjs/melonJS#668]
Ouch! This messes everything up...

All sprites and containers return a relative bounds from me.Rect. Thanks @agmcleod for pointing this out.

**Comment by obiot** [comment]
Ready for merge? :)

**Comment by agmcleod** [comment]
Any ideas on how/when renderable should be updating its bounds? The body does it on entity. But we typically dont invoke renderable's update method anywhere, as it's presently empty. This could definitely break peoples code too. Not sure about calling it every frame performance wise either.

**Comment by obiot** [comment]
if required, probably in the update function (thinking about Spine for example).

else for the most common and basic usage like me.Sprite, or me.AnimationSheet, there is no need ot update the bounds :P

**Comment by agmcleod** [comment]
Well if a sprite is inside a container right. The bounds needs to know its position.

**Comment by obiot** [comment]
Why ? I think sprite should just be display elements, for anything impliying physic/bounds entity should be used, no?

**Comment by agmcleod** [comment]
The main thing is consistency. If _bounds returns absolute for an entity, so should it for any other renderable. The other bit is, one may want to retrieve objects from the quad tree for reasons other than collision.

**Comment by obiot** [comment]
sprite are simple objectc with no logic implementation and no intelligence, so if the end user do manually move the sprite position, he has to use the updateBounds function as well, like he would to with an entity if he also does change the entity pos manually.

**Comment by parasyte** [comment]
It's more about allowing things like `inViewport` checks and `renderHitBox` to work correctly. With nested containers, these things quickly get really ugly if there is any inconsistency.

While working on the center-small-maps bug for #641, I started to play with a container as the PlayerEntity renderable, and realized it was completely no-go. This branch will help, but I'm unsure of what else needs to be done. For sure, the update is a real pain...

**Comment by obiot** [comment]
so if we do agree that the above fix is the correct way to go when modifying manually a renderable position, shall we consider this ticket done ?

**Comment by parasyte** [comment]
Yeah, I guess so... We also talked a bit about automating the bounds cache updating (because it's so very attractive) by creating a Vector2d subclass that can trigger a callback when its x or y properties are updated (using setters). The interface would look like this:

```javascript
this.pos = new me.ObserverVector2d(x, y, {
    "onUpdate" : this.onPosUpdate.bind(this)
});
```

**Comment by obiot** [comment]
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/observe

**Comment by obiot** [comment]
I would rather go for something "standard", but then it's ES6

**Comment by parasyte** [comment]
ES7, actually. :confounded:

**Comment by obiot** [comment]
I propose we merge this one and live with it for the 2.1.0 version, and we can then add a ticket for  a `Object.observe`-like API, either through a polyfill or through a custom similar implementation for a couple of specific object (to be discussed)

**Comment by obiot** [comment]
this said : sport time !

**Comment by parasyte** [comment]
I was looking for some info about what makes `Object.observe` different from a setter. The first thing I found was this benchmark: http://jsperf.com/object-observe-vs-setter (holy crap)

But I didn't find much else in the way of comparisons. Almost everything I see on the Internets about `Object.observe` is how much better it is than polling (no shit)

**Comment by obiot** [comment]
Guys, what do we do with this one?

It's really time to get that 2.1.0 out :)

**Comment by agmcleod** [comment]
I think we should try to do it right the first time, rather than change it later. Don't like the idea of telling our users to do something that'll only last a few months.

**Comment by agmcleod** [comment]
Just to put the discussion here. I did a minor test, and i think the performance loss may not be worth it. Note it's quite the number of invokes im doing, but still: http://jsfiddle.net/6bwuekya/4/. Thinking we should move towards the manual call.

**Comment by obiot** [comment]
to be honest i really do have mix feelings about this one now (as opposed of being against it a couple of days ago).

I'm still though a bit surprise by the benchmark, as even if you did run a lots of iterations, the overhead looks like bigger than what I was actually expecting

**Comment by agmcleod** [comment]
Yeah please look over the code, and let me know if you find any major inefficiencies. It could even just be how i benchmarked it.

**Comment by parasyte** [comment]
I should point out the jsfiddle doesn't actually benchmark a real scenario. Even though it's showing the observable vector is roughly 4x slower, that doesn't account for position vector manipulations not even registering in any game during profiling. :wink:

IMHO, the convenience is absolutely worth 4x on *some* (not all) position vectors.

**Comment by agmcleod** [comment]
What do you mean by some? I think it would apply to most, as this would need to be used for any object that uses a vector in the game world for drawing.

**Comment by parasyte** [comment]
Well, a couple of things come up when searching for vectors that do not need to be observed:

* Entity velocity, acceleration, friction, and collision response
* Points that define shapes
* The `offset` used for TextureAtlas animations
* Scaling factors
* Anchor points
* ImageLayer parallax ratio
* Font sizes
* PointerEvent positions
* Some WebGL stuff...

So yeah, I would say *some* is an appropriate assessment.

**Comment by agmcleod** [comment]
True. I think the number of manual calls to updateBounds() should be similar to the number of observable vectors.

**Comment by parasyte** [comment]
Usually, yes. Except the observable makes it less error-prone, and saves us some trouble with community support; no more forgetting to update the bounding boxes when relative things move.

**Comment by agmcleod** [comment]
Check out my latest commits :)

**Comment by obiot** [comment]
so in the context of observe the object position and update the bounds automatically, it seems that we are done here for 2.1.0 ? :P:P:P

**Comment by parasyte** [comment]
I think the real test will be creating new examples that use nested containers and updating the space invaders tutorial.

Two examples come to mind which will benefit from nested containers:

1. Animation plugins ([Spine](https://github.com/melonjs/melonjs-spine), [Spriter](https://github.com/deangiberson/spriter-scon-js))
2. [Paper dolls](https://github.com/makrohn/Universal-LPC-spritesheet)

**Comment by obiot** [comment]
@parasyte I would rather create separate ticket for them then, and leave it as it for now for this release.

**Comment by obiot** [comment]
I did not test it, but based on what I see, am I right to say that if I call for example any of the other non-redefined function (like `me.ObservableVector2d.scale(x, y)`). the callback will be executed 2 times ? 

Then I would propose to add a private `setNoCheck()` function into `me.Vector2d` (that basically is the `set` one without the exception check`) and update all other function to use thise one instead of directly manipulating `x` and `y`

**Comment by parasyte** [comment]
I'm not sure what you mean by tickets? I'm just saying we need to ensure this code actually works as intended, which comes by way of testing.

Yes, you interpreted the code correctly. Any method on the superclass which assigns both `x` and `y` will run the callback for both. It's not ideal. I would propose to override these functions instead of adding more hacks.

**Comment by obiot** [comment]
just saying that we should maybe release 2.1.0, and then implementing a spiner or paper dolls example in parallel (create a "add spine/paper dolls examples) and eventually fix any potential issue in 2.1.1.

**Comment by parasyte** [comment]
I hope we won't have any issues, actually. :smile: We can ensure that with some simple examples, no?

**Comment by obiot** [comment]
whack-a-mole for example, that's why I reverted the change :P

else as nice improvement, the viewport object could also probably be updated to use the callback mechanism

**Comment by obiot** [comment]
@parasyte about redefining all functions, I was more hoping we could find a way to avoid duplicating code here. Else we could also just update all the original `me.Vector2d` functions to use `set` (since redefined in `me.ObservableVector2d`, but i'm afraid that with both the additional function call and the test on `x` and `y` this would impact performances.

**Comment by parasyte** [comment]
@obiot Just make it a private method, preferably prefixed with underscore. You will be relying on the JIT compiler to be smart enough to inline the function (or else you are actually trading performance for a perceived benefit of compact source code). For example, `scale()`:

Original (makes two calls to `onUpdate` via setters `x` and `y`):

```javascript
scale : function (x, y) {
    this.x *= x;
    this.y *= typeof (y) !== "undefined" ? y : x;
    return this;
},
```

vs. Overridden (single call to `onUpdate`):

```javascript
scale : function (x, y) {
    y = (typeof (y) !== "undefined" ? y : x) * this._y;
    x *= this._x;
    this.onUpdate(x, y, this._x, this._y);
    this._x = x;
    this._y = y;
    return this;
},
```

vs. Hopefully the JIT compiler can inline the `_set` function call (makes one call to `onUpdate` in best case, or two function calls - `_set` and `onUpdate` - in the worst case):

```javascript
_set : function (x, y) {
    this.onUpdate(x, y, this._x, this._y);
    this._x = x;
    this._y = y;
    return this;
},

scale : function (x, y) {
    return this._set(this._x * x, this._y * typeof (y) !== "undefined" ? y : x);
},
```

The last four lines of the "Overridden" version are the common code that would be duplicated between functions. As you can see they have been moved into a private `_set` method for the last version. FTR I am not personally familiar enough with JavaScript JIT inlining rules to make this judgement.

**Comment by obiot** [comment]
me neither, but I just found this article about it :
http://mrale.ph/blog/2012/06/03/explaining-js-vms-in-js-inline-caches.html

**Comment by parasyte** [comment]
It's slower by just a hair: http://jsperf.com/inline-expansion

**Comment by parasyte** [comment]
It's probably a good tradeoff, but I think it won't prevent duplicate code, since `ObservableVector2d` still needs to read the private `_x` and `_y` properties, while `Vector2d` reads its `x` and `y` properties.

**Comment by obiot** [comment]
interesting ! yes, but at least we do not need to reduplicate all the function, just `set` and/or `_set` :)

**Comment by obiot** [comment]
@parasyte see my last commit :)

I don't have the time to do more this afternoon (I only modified the add/sub function to match the current/previous implementation level), so I will call it for now a POC commit, but you can see that with these changes the `me.ObservableVector2d` class is now super small, as only the `_set` function is redefined !

**Comment by parasyte** [comment]
This is what I was talking about in my last comment. Now `add` and `sub` both call the getters. We don't want that. These methods also need to be overridden so they access the private properties instead of the getters.

**Comment by obiot** [comment]
no they don't ? with the `_set` function being redefined (and assuming that me.Vector2d is fully updated, which is not yet the case), nothing is accessing directly `x` and `y` in the ObservableVector class

**Comment by obiot** [comment]
I just added a temporary `console.log` in the callback of the Observable Vector test unit, and just testing the `set`, `add` and `sub` function, the callback is only triggered one time as expected.

**Comment by parasyte** [comment]
@obiot I'm talking about the *getters*. These: https://github.com/melonjs/melonJS/blob/ticket-668/src/math/observable_vector.js#L31 and https://github.com/melonjs/melonJS/blob/ticket-668/src/math/observable_vector.js#L49 Put your `console.log` there.

These getters get called through the superclass, e.g. here: https://github.com/melonjs/melonJS/blob/ticket-668/src/math/vector.js#L104 and here: https://github.com/melonjs/melonJS/blob/ticket-668/src/math/vector.js#L116

This is why I suggested overriding all of the methods, and not relying on the code in the superclass.

Unless you want to add private `_x` and `_y` properties to `me.Vector2d` as well, I don't think you can eliminate code duplication while remaining fast.

**Comment by obiot** [comment]
The getter.... I was so not thinking about the getter and only thinking about the setter :(

**Comment by obiot** [comment]
I really hate duplicating code :( this said, maybe we can still modify also me.Vector2d to use the private set function in order to keep them both as "close as possible" (i'd suppose it would be less error-prone then)

**Comment by obiot** [comment]
@parasyte finish updating everything, but i'm having some weird exception, and cannot futher debug it... would you be up for it if i commit the broken code? :P

**Comment by obiot** [comment]
also, i'm not sure what to do with functions that returns an actual copy of the object, shall we return a `me.Vector` one or and `me.ObservableVector2d` one with the callback also duplicated ?

**Comment by obiot** [comment]
kind of f**** up the commit with a double-revert, but here is what I've done so far. Spent some time debugging it, but i won't be able to do more for today, if any living soul out-there can help finishing it that would be great :P

**Comment by obiot** [comment]
and of course, the branch is currently broken, the platformer does not work anymore !

**Comment by parasyte** [comment]
The function contract is to return a reference to itself, so `return this`. The `clone` method should return a new object which is of the same class; `me.Vector2d` returns `new me.Vector2d()`, and `me.ObservableVector2d` returns `new me.ObservableVector2d()` ... (ouch, these names are hard to type)

I can take a look at it when I get home.

**Comment by obiot** [comment]
@parasyte, fully agree with that, but my concern is that assuming you clone the position vector for some temporary mathematical operation or whatever, you end up with another `new me.ObservableVector2d()` object also triggering your callback.

**Comment by agmcleod** [comment]
I think that's what copy is for right. They use copy to return a new vector? Perhaps copy needs to accept an options/settings parameter for observablevector

**Comment by parasyte** [comment]
Clone should be a clone, by design, no? Maybe it's not useful, but it is semantically correct that a clone would have the same update callback by default. I would agree with @agmcleod's suggestion, and offer this alternative:

`var clone = this.pos._super(me.Vector2d, "clone");`

**Comment by obiot** [comment]
oh boy, it feels like we are going to spend another month on this one :P:P:P

**Comment by obiot** [comment]
all the following functions are returning a new instance object :
* `clamp` : https://github.com/melonjs/melonJS/blob/ticket-668/src/math/observable_vector.js#L161
* `floor` : https://github.com/melonjs/melonJS/blob/ticket-668/src/math/observable_vector.js#L161
* `ceil` : https://github.com/melonjs/melonJS/blob/ticket-668/src/math/observable_vector.js#L231
* `negate` : https://github.com/melonjs/melonJS/blob/ticket-668/src/math/observable_vector.js#L253
* `clone` : https://github.com/melonjs/melonJS/blob/ticket-668/src/math/observable_vector.js#L441

`copy` does not return a new object, it just copy the value from a vector to another one

**Comment by obiot** [comment]
all fixed and working now !

now only a few meters left before the arrival line :
- remove calls to `updateBounds()` : not only in examples, but as well in the engine itself, as when running the platformer example I still noticed a couple of console warn on it.
- ensure all examples are working
- maybe a `toVector2d()` function to `me.ObservableVector2d()` as a clean way to return a vector object out of it ? (if required for math operation without the need of the callback being duplicated-triplicated)

I'm done for the day, so if you guys have some time today, this is your to-do list :):):)

**Comment by obiot** [comment]
@parasyte i just checked the collision example, removed the (now) duplicated calls to `updateBounds` and notice that not only it runs slower than before, but as well the behavior of the smileys when colliding with each other is kind of weird. Could you have a look to see what you think could be the cause. I thought it was initially due to the added unnecessary calls to `updateBounds` but removing them did not change that much to it

**Comment by agmcleod** [comment]
For `updateBounds` last calls are mostly in body. Not sure if it needs to be removed or not. Removed the call from the draggable entity.

**Comment by agmcleod** [comment]
Tested examples on my machine, seem good. This might be too quick to notice the perf issues you refer to in collision test.

**Comment by parasyte** [comment]
I added a deprecation warning, so it should be very easy to tell if the method is still being called. I haven't looked at anything in a few days. Getting over a cold and stuff. ;(

**Comment by parasyte** [comment]
Quick profiler results, comparing master to ticket-668:

* `me.Entity.update` : 0.85% -> 3.2%
* `me.ObservableVector2d._set` : N/A -> 2.06%
* `me.Entity.draw` : 0.26% -> 1.97%
* `me.Sprite.draw` : 0.43% -> 1.82%
* `me.Container.update` : 0.25% -> 1.33%
* `me.Entity.updateBoundsPos` N/A ->  0.87%

This is actually a huge difference... An order of magnitude performance loss in these methods.

**Comment by parasyte** [comment]
Oh jeez, why does every renderable use the observable vector? I thought we only wanted that for Entity?

**Comment by agmcleod** [comment]
@parasyte that was the point of the whole ticket. To make it so a renderable updates its position each frame similar to how entity was.

**Comment by parasyte** [comment]
The last commit just fixes the shapes in collision_test, it already helps a little bit. We need to fix the Renderable position, though. Right now we are incurring a performance hit on the position getters for every single draw.

@agmcleod `me.Entity` needs an observable position so it can update its bounds. Not every renderable has a disconnected bounds. `me.Sprite` for example should reference itself as its bounds, this is functionality that it inherits from `me.Rect`. There's nothing dynamic about a simple sprite that requires such complexity.

**Comment by agmcleod** [comment]
It's the absolute positioning that's the problem. The main reason I implemented this was to solve that. Nothing about containing a separate bounds, but to update its absolute position.

**Comment by parasyte** [comment]
Aw, crap. For some reason I thought the container was still doing that. :confounded:

**Comment by obiot** [comment]
while thinking how to improve things a bit, I stumbled on this test :
http://jsperf.com/function-calls-direct-vs-apply-vs-call-vs-bind/67

Interestingly , apply seems to be super slow compared to call, and we have a bunch of interesting places using it (container update , AnimationSheet update. entity update, updateBoundsPos)

**Comment by agmcleod** [comment]
Inheritance needs apply, because it passes an array of arguments, instead of single arguments.

**Comment by parasyte** [comment]
There was another quick chat in gitter about this, basically it boils down to a need to conditionally enable the observable position vector. Only direct descendants of `me.Container` need the observable position vector, so we might just revert the default for `me.Renderable` to use a standard vector, and add a method that allows "upgrading" it to an observable one. Finally, `me.Container.addChild` can call the "upgrade" method if the child is a renderable.

**Comment by obiot** [comment]
@agmcleod sorry i'm a bit confused with the code changes, but is the `_absPos` stuff in the container still  used ?

**Comment by obiot** [comment]
@parasyte, so to rephrase with my own words what you say, if a objet is a direct descendant of a container, it's position needs to be a `me.ObservableVector2d`, else we just need a `me.Vector2d` object.

which means that for example :
* an entity added to a container -> `me.ObservableVector2d`
* an renderable added to a container -> `me.ObservableVector2d`
* an renderable added to an entity -> `me.Vector2d`

In that case I would put the code responsible of changing an object pos to an obersvable one directly in the container `addChild()` function, I think it makes sense.

**Comment by obiot** [comment]
so I did the following test :

- replaced the following instantiation by using a "standard" `me.Vector2d()` object 
https://github.com/melonjs/melonJS/blob/ticket-668/src/renderable/renderable.js#L156

- added the following code in both the `addChild` and `addChildAt` container function (here https://github.com/melonjs/melonJS/blob/ticket-668/src/renderable/container.js#L153 and here https://github.com/melonjs/melonJS/blob/ticket-668/src/renderable/container.js#L153 ) :
````javascript
// convert the object child position
// to a corresponding observable vector
if (child.pos instanceof me.Vector2d) {
    child.pos = new me.ObservableVector2d(
        child.pos.x,
        child.pos.y,
        {onUpdate: child.updateBoundsPos.bind(child)}
    );
}
````

seems to be working for almost everything, but when running the platformer, the level scrolling is not happening... i'm not super sure why :P

**Comment by parasyte** [comment]
Because viewport is following the position vector that you are destroying in `addChild`

**Comment by parasyte** [comment]
Move this line: https://github.com/melonjs/melonJS/blob/ticket-668/src/renderable/renderable.js#L156 to `me.Entity.init` and use with your current patch. That will fix the scrolling/viewport follow issue.

**Comment by obiot** [comment]
I wish we would have a way to "mutate" one object A into an object B by keeping the same global references (like a nice cast in C/C++) !

I'll try with that fix tomorrow morning, running out of time for now here as I have to run to catch my bus :)

**Comment by agmcleod** [comment]
@obiot you bring up a good point. Do we just need to reference _bounds.pos instead of the _absPos?

**Comment by obiot** [comment]
That's what I initiallt thought as well, but i'm not sure i'm familiar enough with all the changes linked to this ticket :p

**Comment by agmcleod** [comment]
@terebentina brought up a good point in Gitter. Moving a container, until the set x or y setter is invoked, the _bounds won't be updated will it?

**Comment by parasyte** [comment]
@agmcleod That's exactly what all this work was supposed to solve.

**Comment by obiot** [comment]
@parasyte not getting it working by adding this, there is something else :P also it starts to look like a big hack rather than something clean and more efficient, so i'm really wondering if it worth investigating that particular one.

**Comment by obiot** [comment]
@agmcleod that one is on you :P:P:P

**Comment by obiot** [comment]
@agmcleod i'm confused as well when it comes to the container, as then we have both `_bounds` and `childBounds` ?

**Comment by obiot** [comment]
i just tried to replace all occurences of `_absPos` by `_bounds.pos`, but all the collision stuff is then screwed up, and I'm still too much confused between `_absPos`, `_bounds` and `childBounds` to really understand what I'm doing, so.... stopping here, going back home !

else tomorrow is April 1st :)

**Comment by parasyte** [comment]
@obiot You have valid concerns. The observer pattern is really not supposed to make it more efficient, but it does hide some (all?) of the cache invalidations. From a user perspective, that has some importance. Anyone using the framework should not be concerned with its internals; Invalidating the boundary box cache when an item position changes is such a concern. It is better for the framework to be smart about cache invalidation (this ticket) or not cache at all. Either option removes the complexity of cache invalidation from user-side code.

For your other questions:

* `_absPos` is the absolute position for a *parent*. The object itself never reads this property; `me.Container` will set its own `_absPos` and its children directly read the property. It's kind of a hack that the property exists on `me.Renderable` instead of `me.Container`, because some non-container renderables can also be a parent (`me.Entity` et al.)
* `_bounds.pos` is the absolute position; it is the sum of the relative position + the parent's `_absPos`.
* `childBounds` is a silly thing and I don't know if there is a use for it? It provides a minimal bounding box for a container's children.

**Comment by obiot** [comment]
Ok guys, i'm leaving (again) for a small trip, when i come back (Sunday), i will merge and release 2.1.0, seems like we don't really have a solution to improve things in the short time, and better have something than nothing (not to mention that the performance drop is not that significant and should even be not visible for most of the games).

**Comment by agmcleod** [comment]
Think we should merge this soon, or any low hanging fruit with it still? Honestly once we get this out in the wild, feedback from others may give us some ideas. What bits of it are confusing to users? When we explain things in the support forum, does it raise other questions? And things of that nature.

**Comment by obiot** [comment]
indeed we should :P

**Comment by agmcleod** [comment]
And merged.
